### PR TITLE
User can read articles from external news agencies  

### PR DESCRIPTION
**Pull request by holgertidemand**
#### Description
Changes proposed in this pull request:

* Displays News articles from various external sources 
#### What I have learned working on this feature:
* How to test fetching from an API 
* How to use News-api 

#### Screenshots:
screen shot 2018-03-18 at 13 16 59

### CODE REVIEW

File: app/controllers/index_controller.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
+require 'news-api'
 class IndexController < ApplicationController
   def index
+    api_key = News.new("e9b2cff169de4096a9d34df4e3742ab2")
```
Comment (tochman) [inline app/controllers/index_controller.rb]: Oh! API key in VCS? Really?

**Review by tochman — CHANGES_REQUESTED** [review]

File: app/controllers/index_controller.rb (hunk 1 above)
Comment (holgertidemand) [reply app/controllers/index_controller.rb]: Can you please tell me how to make this code work now when we removed the key from VCS. @tochman @diraulo

File: features/fetching_articles_from_news_api.feature [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Feature: Website should contain news from API
+  As a dev team
```
Comment (tochman) [inline features/fetching_articles_from_news_api.feature]: Please clarify what kind of stakeholder “dev team” is.

File: config/application.rb [hunk 3]
```diff
@@ -29,5 +29,15 @@ class Application < Rails::Application
...
+
+    config_files = ['secret.yml']
+
+    config_files.each do |file_name|
+      file_path = File.join(Rails.root, 'config', file_name)
+      config_keys = HashWithIndifferentAccess.new(YAML::load(IO.read(file_path)))[Rails.env]
+      config_keys.each do |k,v|
+        ENV[k.upcase] ||= v
+      end
+    end
```
Comment (tochman) [inline config/application.rb]: Please explain what this code is doing for you.

File: app/controllers/index_controller.rb [hunk 4]
```diff
@@ -1,4 +1,7 @@
+require 'news-api'
 class IndexController < ApplicationController
   def index
+    api_key = News.new(ENV["SECRET_KEY_NEWS_API"])
+    @api_new_feed = api_key.get_top_headlines(category: 'technology', country: 'us', pageSize: 5)
```
Comment (tochman) [inline app/controllers/index_controller.rb]: Is it a ‘new_feed’ or ‘news_feed’?

File: Gemfile [hunk 5]
```diff
@@ -5,6 +5,8 @@ ruby '2.4.1'
 
 gem 'rails', '~> 5.2.0.rc1'
 gem 'pg', '>= 0.18', '< 2.0'
+gem 'dotenv-rails'
```
Comment (tochman) [inline Gemfile]: Is this gem/library used in the solution you chose?

**Comment by tochman** [comment]
This need to be reopen and comments addressed. Do not create new PR when facing dependency issues. Work through them instead.

**Comment by tochman** [comment]
Who of the coaches was assisting you before you decided to close this one?

File: Gemfile (hunk 5 above)
Comment (holgertidemand) [reply Gemfile]: Yes.

File: app/controllers/index_controller.rb (hunk 4 above)
Comment (holgertidemand) [reply app/controllers/index_controller.rb]: news.

File: config/application.rb (hunk 3 above)
Comment (holgertidemand) [reply config/application.rb]: Nothing! Part of the solution to remove it

File: app/controllers/index_controller.rb (hunk 4 above)
Comment (tochman) [reply app/controllers/index_controller.rb]: So why does it say ‘new’

File: features/fetching_articles_from_news_api.feature (hunk 2 above)
Comment (holgertidemand) [reply features/fetching_articles_from_news_api.feature]: Will change this.

File: Gemfile (hunk 5 above)
Comment (tochman) [reply Gemfile]: So what is the rest of the changes introduced in this PR actually do?

**Comment by diraulo** [comment]
Semaphore is complaining here about the `MASTER_KEY` not being present in the CI environment. Can someone look into that?

**Comment by tochman** [comment]
The salt in `config/master.key` needs to be added to the CI's env variables as `RAILS_MASTER_KEY`

File: Gemfile [hunk 6]
```diff
@@ -14,6 +15,7 @@ gem 'devise'
 
 
 group :development, :test do
+  gem 'dotenv-rails'
```
Comment (magnus-thor) [inline Gemfile]: You dont need this gem if you are using rails credentials.

File: app/views/index/index.html.erb [hunk 7]
```diff
@@ -7,3 +7,11 @@
   <li><%= link_to 'Login', new_user_session_path %></li>
   <li><%= link_to 'Register', new_user_registration_path %></li>
 <% end %>
+
+<% @api_news_feed.each do |article|%>
+<article id="news">
+  <h1><%=article.title %></h1>
+  <p><%=article.description %></p>
+  <%= image_tag(article.urlToImage, size: '150x100', alt: "Source not found") %>
+</article>
```
Comment (magnus-thor) [inline app/views/index/index.html.erb]: fix the indentation

File: features/user_can_submit_article_for_review.feature [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Feature: User should be able to sumbit article for review
+  As a journalist
+  In order to be able to get my articles published
+  I would like to be able to submit the articles for review
```
Comment (magnus-thor) [inline features/user_can_submit_article_for_review.feature]: Is this file needed for this PR?

**Review by magnus-thor — DISMISSED** [review]

File: Gemfile (hunk 6 above)
Comment (holgertidemand) [reply Gemfile]: Well this is what made it work for us!

File: features/user_can_submit_article_for_review.feature (hunk 8 above)
Comment (holgertidemand) [reply features/user_can_submit_article_for_review.feature]: No. will remove this

File: Gemfile (hunk 6 above)
Comment (tochman) [reply Gemfile]: They are NOT using encrypted credentials @magnus-thor

**Review by tochman — DISMISSED** [review]

File: app/controllers/index_controller.rb [hunk 9]
```diff
@@ -1,4 +1,7 @@
+require 'news-api'
 class IndexController < ApplicationController
   def index
+    api_key = News.new(Rails.application.credentials.news[:news_api_key])
```
Comment (diraulo) [inline app/controllers/index_controller.rb]: The variable naming here is a bit off...

You're initializing the News API client here so variable names like `news_api_client`, `news_client` are probably better suited than `api_key`. You're not storing the API key in that variable

File: config/environments/production.rb [hunk 10]
```diff
@@ -3,6 +3,7 @@
 
   # Code is not reloaded between requests.
   config.cache_classes = true
+
```
Comment (diraulo) [inline config/environments/production.rb]: Remove empty line on this file

File: features/fetching_articles_from_news_api.feature [hunk 11]
```diff
@@ -0,0 +1,9 @@
+Feature: Website should contain news from API
+  As a owner
+  In order to provide more articles for users to read
+  We would like to pull articles from other sources using an API
+
+
+Scenario: User visit the Index page
+  Given I am on the "Index" page
+  Then I should see 5 articles
```
Comment (diraulo) [inline features/fetching_articles_from_news_api.feature:9]: One thing with this test is that you're making an actual call to the real API during the test to retrieve articles. This will have an impact on the performance of this test. It'll be slow because of the network call but also if the API doesn't work then your tests will fail. 

One way to deal with this is to create stubs for network requests using a gem like [WebMock](https://github.com/bblimke/webmock). You can either deal with this now on this PR (if you chose to) or create a ticket to address this separately.

**Review by diraulo — COMMENTED** [review]
See comments... Some should be addressed before I approve this PR, others you'll decide whether to fix them here or not

File: features/fetching_articles_from_news_api.feature (hunk 11 above)
Comment (jyuw) [reply features/fetching_articles_from_news_api.feature:9]: Should we merge as is first and then implement this as a new PR to enhance the API feature or would you like us to rewrite this story/tasks/tests?

Comment (diraulo) [reply features/fetching_articles_from_news_api.feature:9]: You can merge if you've fixed the other issues I've mentioned above

**Review by diraulo — APPROVED** [review]

**Review by larathordar — APPROVED** [review]
Looks good !
